### PR TITLE
Integrate Mender

### PR DESCRIPTION
**Pull request by eugulixes**

### CODE REVIEW

File: essentials.sh [hunk 1]
```diff
@@ -17,6 +17,18 @@ ALPINE_VER="3.7"
...
 
+MENDER_ARTIFACT_VER="2.3.0"
+
+MENDER_CLIENT_VER="1.6.0"
+
+MENDER_CLIENT_REPO="https://raw.githubusercontent.com/mendersoftware/mender"
+
+MENDER_CLIENT_REVISION="1.6.x"
+
+META_MENDER_REPO="https://raw.githubusercontent.com/mendersoftware/meta-mender"
```
Comment (denismosolov) [inline essentials.sh]: Not used
```suggestion
```

File: essentials.sh [hunk 2]
```diff
@@ -17,6 +17,18 @@ ALPINE_VER="3.7"
...
+
+META_MENDER_REVISION="sumo"
```
Comment (denismosolov) [inline essentials.sh]:
```suggestion
```

File: essentials.sh [hunk 3]
```diff
@@ -33,6 +49,14 @@ text_in_yellow_color=$(tput setaf 3)
 
 reset=$(tput sgr0)
 
+# Images types
+
+IMAGE_FOR_RPI=1
```
Comment (denismosolov) [inline essentials.sh]: How about "PLAIN_IMAGE"

File: essentials.sh [hunk 4]
```diff
@@ -33,6 +49,14 @@ text_in_yellow_color=$(tput setaf 3)
...
+
+IMAGE_FOR_RPI_WITH_MENDER_CLIENT=2
```
Comment (denismosolov) [inline essentials.sh]: Could be "IMAGE_WITH_MENDER"

File: essentials.sh [hunk 5]
```diff
@@ -33,6 +49,14 @@ text_in_yellow_color=$(tput setaf 3)
...
+
+IMAGE_ARTIFACT=3
```
Comment (denismosolov) [inline essentials.sh]: How about "MENDER_IMAGE_ARTIFACT"?

File: essentials.sh (hunk 4 above)
Comment (denismosolov) [reply essentials.sh]: Or MENDER_COMPATIBLE_IMAGE

File: README.md [hunk 6]
```diff
@@ -51,8 +52,17 @@ Pieman is a script for creating custom OS images for single-board computers such
...
+  * bzip2
+  * xz
+* To enable [Mender](https://mender.io) support:
+  * Development libraries and header files related to C standard library (make sure the package, which is going to be installed to satisfy the dependency, includes `/usr/include/sys/types.h`)
+  * C programming language compiler
+  * Go programming language compiler
+  * bc
+  * dtc
+  * make
+  * mkimage
```
Comment (denismosolov) [inline README.md]: Here is the list of packages for Ubuntu 18.04 if you want to add them
`bc gcc device-tree-compiler golang-go make u-boot-tools libc6-dev-i386`

File: pieman.sh [hunk 7]
```diff
@@ -246,4 +337,8 @@ else
     image="${IMAGE}${extension}"
 fi
 
-success "${image} was built. Use Etcher (https://etcher.io) to burn it to your SD card."
+if ${CREATE_ONLY_ARTIFACT}; then
```
Comment (denismosolov) [inline pieman.sh]: Let's use present tense here. Similar to `Chroot environment ${R} is done.` for chroot,

**Comment by denismosolov** [comment]
Let's stop the script if Mender dependencies aren't satisfied. How about adding extra checking for `mender_dependencies_are_satisfied`, `ENABLE_MENDER` and `CREATE_ONLY_ARTIFACT`

File: README.md [hunk 8]
```diff
@@ -378,6 +388,40 @@ Note, that the parameter must follow the format "uid:gid" where `uid` and `gid`
 
 Allows specifying the Python 3 interpreter to be used when calling the Pieman-specific utilities. 
 
+#### Extra
+
+##### CREATE_ONLY_ARTIFACT=false
```
Comment (denismosolov) [inline README.md]: How about `MENDER_CREATE_ONLY_ARTIFACT`?

File: README.md [hunk 9]
```diff
@@ -378,6 +388,40 @@ Note, that the parameter must follow the format "uid:gid" where `uid` and `gid`
...
+
+Specifies whether to install the Mender client to provide for OTA updates.
+
+Note that the OTA updates support is currently limited to 32-bit Raspbian and Ubuntu for Raspberry Pi 3 Model B.
+
+##### MENDER_ARTIFACT_NAME="release-1_1.6.0"
+
+...
+
+##### MENDER_DATA_SIZE=128
```
Comment (denismosolov) [inline README.md]: Would be great if we create a parameter which will extend the last partition if set to true and don't extend if set to to false.
But for now let's add a note that mender data partition will be extended on the first boot.

File: bootstrap/12-pm.sh [hunk 10]
```diff
@@ -25,6 +25,13 @@ if ${ALLOW_UNAUTHENTICATED}; then
     fi
 fi
 
+if ${ENABLE_MENDER}; then
+    add_package_to_includes ca-certificates
+    # Need to have the reboot executable file in $PATH
+    add_package_to_includes systemd-sysv
+    add_package_to_includes u-boot-tools
```
Comment (denismosolov) [inline bootstrap/12-pm.sh]: Not sure if we need u-boot-tools

File: docker/Dockerfile [hunk 11]
```diff
@@ -36,6 +36,21 @@ RUN touch .dockerenv \
     wget \
     whois \
     xz \
+    # The following packages are necessary for building Das U-Boot and Mender
```
Comment (denismosolov) [inline docker/Dockerfile]: I wonder if u-boot-tools is missed?

File: README.md (hunk 8 above)
Comment (eugulixes) [reply README.md]: I wanted it to look like another similar param -- `CREATE_ONLY_CHROOT`.

File: essentials.sh [hunk 12]
```diff
@@ -107,6 +131,28 @@ def_var() {
...
+# Acts like def_var but places further restrictions on the target value: it
+# must be integer.
+# Globals:
+#     None
+# Arguments:
+#     Variable name
+#     Value by default
+# Returns:
+#     None
+def_int_var() {
```
Comment (denismosolov) [inline essentials.sh]: Should we add this to test_defining_variables?

File: README.md (hunk 9 above)
Comment (eugulixes) [reply README.md]: I created a corresponding issue https://github.com/tolstoyevsky/pieman/issues/124.

File: essentials.sh (hunk 3 above)
Comment (eugulixes) [reply essentials.sh]: I wanted all the params to start with the `IMAGE_` prefix.

File: pieman.sh (hunk 7 above)
Comment (eugulixes) [reply pieman.sh]: There is a corresponding issue https://github.com/tolstoyevsky/pieman/issues/122.

File: README.md (hunk 8 above)
Comment (eugulixes) [reply README.md]: Renamed it into `CREATE_ONLY_MENDER_ARTIFACT`.

File: essentials.sh (hunk 5 above)
Comment (denismosolov) [reply essentials.sh]: Or `IMAGE_MENDER_ARTIFACT`?

File: docker/Dockerfile (hunk 11 above)
Comment (eugulixes) [reply docker/Dockerfile]: You're right, but in Alpine it's called `uboot-tools`.

File: files/mender/mender.conf.template [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
+    "InventoryPollIntervalSeconds": 5,
```
Comment (denismosolov) [inline files/mender/mender.conf.template]: We'll need parameters to specify poll interval
https://docs.mender.io/1.7/client-configuration/configuration-file/polling-intervals#how-to-choose-right-intervals

File: docker/docker-entrypoint.sh [hunk 14]
```diff
@@ -8,4 +8,4 @@ cd /root/pieman && ./pieman.sh
 # some platforms:
 # E: Cannot install into target '/build/<project name>/chroot' mounted with
 # noexec or nodev
-mv $(ls /root/pieman/build/*.img{.bz2,.gz,.xz,} 2> /dev/null) /result
+mv $(ls /root/pieman/build/*.{img,img.bz2,img.gz,img.xz,mender} 2> /dev/null) /result
```
Comment (denismosolov) [inline docker/docker-entrypoint.sh]: It can be mender.bz2, mender.gz or mender.xz

File: files/mender/device_type [hunk 15]
```diff
@@ -0,0 +1 @@
+device_type=raspberrypi3
```
Comment (denismosolov) [inline files/mender/device_type]: Any ideas how to move that to devices directory?

File: pieman.sh [hunk 16]
```diff
@@ -183,48 +212,114 @@ run_scripts "bootstrap"
...
+
+    umount "${MOUNT_POINT}"
+
+    info "converting ${IMAGE} into artifact"
+
+    "${TOOLSET_DIR}"/mender/mender-artifact write rootfs-image \
+        --update "${IMAGE}" \
+        --output-path "${ARTIFACT}" \
+        --artifact-name "${MENDER_ARTIFACT_NAME}" \
+        --device-type raspberrypi3
```
Comment (denismosolov) [inline pieman.sh]: What if we create a 'mapper' which will map our `DEVICE` parameter into this value? The same solution for the files/mender/device_type file

File: files/mender/mender.conf.template (hunk 13 above)
Comment (denismosolov) [reply files/mender/mender.conf.template]: Or even a set of parameters to provide all possible key-value pairs in mender.conf

Comment (mirzak) [reply files/mender/mender.conf.template]: Just wanted to highlight that we changed these in the mender-convert tool as well, https://github.com/mendersoftware/mender-convert/commit/366cf37453e8728a6c68e622b0428f1ecfcc5fec.

Also wanted to point that that these are "demo" values only intended to be used together with the [Demo environment](https://docs.mender.io/1.6/getting-started/create-a-test-environment).

We default to these values in a "production" build, 

MENDER_UPDATE_POLL_INTERVAL_SECONDS ?= "1800"
MENDER_INVENTORY_POLL_INTERVAL_SECONDS ?= "1800"
MENDER_RETRY_POLL_INTERVAL_SECONDS ?= "300"

File: files/mender/version [hunk 17]
```diff
@@ -0,0 +1 @@
+2
```
[the remark this answers is not in the document]
Comment (mirzak) [reply files/mender/version]: This should be installed to `/etc/mender/scripts/version` and relates to [state-scripts](https://docs.mender.io/1.6/artifacts/state-scripts) and is a why of saying which "version format" of state-scripts is supported.

It is only needed if  [state-scripts](https://docs.mender.io/1.6/artifacts/state-scripts) are used.

File: helpers/base.sh [hunk 18]
```diff
@@ -314,45 +314,103 @@ cleanup() {
...
-# partition and the second one is the partition intended for the root
-# filesystem.
+# Creates an image with the specified number of partitions.
+# TODO: describe labels.
 # Globals:
 #     IMAGE
 # Arguments:
-#     Root filesystem size in bytes
+#     Alignment size in megabytes
+#     Partition1 size in megabytes
```
Comment (denismosolov) [inline helpers/base.sh]: It supposed to be called fstype:size
Let's put a note about fs type and default value for fs type

**Comment by eugulixes** [comment]
@denismosolov, the output is ok, since the Docker image has already got the toolset. So, it's not necessary to build or rebuild it and it's not necessary to keep all the Mender dependencies for that.

File: files/mender/version (hunk 17 above)
[the remark this answers is not in the document]
Comment (denismosolov) [reply files/mender/version]: @eugulixes we don't provide state scripts and I suggest to remove that file and remove related parameters from mender.conf once merged. What do you think?

File: files/mender/mender.conf.template [hunk 19]
```diff
@@ -0,0 +1,9 @@
...
+    "RetryPollIntervalSeconds": 30,
+    "RootfsPartA": "/dev/mmcblk0p2",
+    "RootfsPartB": "/dev/mmcblk0p3",
+    "ServerURL": "{SERVER_URL}",
+    "TenantToken": "{TENANT_TOKEN}",
```
Comment (denismosolov) [inline files/mender/mender.conf.template]:
```suggestion
    "TenantToken": "{MENDER_TENANT_TOKEN}",
```

File: helpers/mender.sh [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+    #
+    # Install Mender client
+    #
+
+    # Install the daemon executable and its configuration files
+    install -m 0755 "${TOOLSET_DIR}/mender/mender" "${MOUNT_POINT}"/usr/bin/mender
+    install -d "${etc_mender_dir}"
+    install -m 0644 "${PIEMAN_DIR}"/files/mender/mender.conf.template "${etc_mender_dir}"/mender.conf
+    sed -i -e "s#{SERVER_URL}#${MENDER_SERVER_URL}#" "${etc_mender_dir}"/mender.conf
+    sed -i -e "s#{TENANT_TOKEN}#${MENDER_TENANT_TOKEN}#" "${etc_mender_dir}"/mender.conf
```
Comment (denismosolov) [inline helpers/mender.sh]:
```suggestion
    sed -i -e "s#{MENDER_TENANT_TOKEN}#${MENDER_TENANT_TOKEN}#" "${etc_mender_dir}"/mender.conf
```

File: files/mender/mender.conf.template [hunk 21]
```diff
@@ -0,0 +1,9 @@
+{
+    "InventoryPollIntervalSeconds": 5,
+    "RetryPollIntervalSeconds": 30,
+    "RootfsPartA": "/dev/mmcblk0p2",
+    "RootfsPartB": "/dev/mmcblk0p3",
+    "ServerURL": "{SERVER_URL}",
```
Comment (denismosolov) [inline files/mender/mender.conf.template]:
```suggestion
    "ServerURL": "{MENDER_SERVER_URL}",
```

File: helpers/mender.sh [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+
+    #
+    # Install Mender client
+    #
+
+    # Install the daemon executable and its configuration files
+    install -m 0755 "${TOOLSET_DIR}/mender/mender" "${MOUNT_POINT}"/usr/bin/mender
+    install -d "${etc_mender_dir}"
+    install -m 0644 "${PIEMAN_DIR}"/files/mender/mender.conf.template "${etc_mender_dir}"/mender.conf
+    sed -i -e "s#{SERVER_URL}#${MENDER_SERVER_URL}#" "${etc_mender_dir}"/mender.conf
```
Comment (denismosolov) [inline helpers/mender.sh]:
```suggestion
    sed -i -e "s#{MENDER_SERVER_URL}#${MENDER_SERVER_URL}#" "${etc_mender_dir}"/mender.conf
```

File: files/mender/mender.conf.template (hunk 13 above)
Comment (denismosolov) [reply files/mender/mender.conf.template]: re #129

File: files/mender/mender.service [hunk 23]
```diff
@@ -0,0 +1,15 @@
+[Unit]                     
+Description=Mender OTA update service                         
+After=systemd-resolved.service                                
+
+[Service]                  
+Type=idle                  
+User=root                  
+Group=root                 
+ExecStartPre=/bin/mkdir -p -m 0700 /data/mender               
+ExecStartPre=/bin/ln -sf /etc/mender/tenant.conf /var/lib/mender/authtentoken
```
Comment (denismosolov) [inline files/mender/mender.service]: Hi @mirzak 
Can you tell us if this line is necessary?
I can see a note in this article "Editor: the tenant.conf is only needed if you are using the Mender hosted service"
https://mender.io/blog/notes-from-a-user-on-mender-without-yocto
But I cannot find any reference to tenant.conf in https://docs.mender.io/1.6

File: files/mender/version (hunk 17 above)
[the remark this answers is not in the document]
Comment (denismosolov) [reply files/mender/version]: Let's postpone it. Image doesn't boot without version file

File: helpers/base.sh [hunk 24]
```diff
@@ -314,45 +314,112 @@ cleanup() {
...
 
-    dd if=/dev/zero of="${IMAGE}" bs=1 seek=${image_size} count=1
+    latest="$((${#fs_types[@]} - 1))"
+    start="${alignment}"
+    for i in $(seq 0 "${latest}"); do
+        end="$((start + partition_sizes[i]))"
 
-    parted "${IMAGE}" mktable msdos
+        if [ "${i}" == "${latest}" ]; then
+            parted -s "${IMAGE}" -- mkpart primary ext2 "${start}MiB" -1s
```
Comment (denismosolov) [inline helpers/base.sh]:
```suggestion
            parted -s "${IMAGE}" -- mkpart primary "${fs_types[i]}" "${start}MiB" -1s
```

File: helpers/base.sh [hunk 25]
```diff
@@ -314,45 +314,112 @@ cleanup() {
...
+#     None
+format_partitions() {
+    local args=("$@")
+    local mkfs=""
+
+    for i in $(eval echo "{1..$#}"); do
+        mkfs="mkfs.${args[i - 1]}"
+        case "${args[i - 1]}" in
+        fat|vfat)
+            ${mkfs} -n boot -F 32 -v "${LOOP_DEV}p${i}" 1>&2-
```
Comment (denismosolov) [inline helpers/base.sh]: It set the label 'boot' to any fat partition

Comment (denismosolov) [reply helpers/base.sh]: re #130

File: helpers/fs.sh [hunk 26]
```diff
@@ -19,15 +19,18 @@
 #     PYTHON
 # Arguments:
 #     Directory name
+#     Optiones to be passed to du.py
```
Comment (denismosolov) [inline helpers/fs.sh]:
```suggestion
#     Options to be passed to du.py
```

File: files/mender/device_type (hunk 15 above)
Comment (eugulixes) [reply files/mender/device_type]: I made a note about it. I suggest we will discuss it later.

File: pieman.sh [hunk 27]
```diff
@@ -183,48 +228,114 @@ run_scripts "bootstrap"
...
+    install_mender
+
+    umount "${MOUNT_POINT}"
+
+    # The data partition has to contain /mender/device_type and
+    # /u-boot/fw_env.config
+    mount "${LOOP_DEV}p4" "${MOUNT_POINT}"
+    mkdir "${MOUNT_POINT}"/mender
+    mkdir "${MOUNT_POINT}"/u-boot
+    install -m 0644 "${PIEMAN_DIR}"/files/mender/device_type "${MOUNT_POINT}"/mender
```
Comment (denismosolov) [inline pieman.sh]:
```suggestion
    install -m 0444 "${PIEMAN_DIR}"/files/mender/device_type "${MOUNT_POINT}"/mender
```

File: toolset.sh [hunk 28]
```diff
@@ -57,6 +72,83 @@ else
...
+    pushd "${TOOLSET_DIR}/mender/uboot-mender"
+        info "building Das U-Boot"
+
+        ARCH=arm CROSS_COMPILE="${cross_compiler}" make --quiet distclean
+        ARCH=arm CROSS_COMPILE="${cross_compiler}" make rpi_3_32b_defconfig
+        ARCH=arm CROSS_COMPILE="${cross_compiler}" make
+        ARCH=arm CROSS_COMPILE="${cross_compiler}" make envtools
+
+        cp "u-boot.bin" "${TOOLSET_DIR}/mender"
+        cp tools/env/fw_printenv "${TOOLSET_DIR}/mender"
```
Comment (denismosolov) [inline toolset.sh]: Device specific

File: toolset.sh [hunk 29]
```diff
@@ -57,6 +72,83 @@ else
...
+        info "fetching Mender Artifacts Library from https://github.com/mendersoftware/mender-artifact.git"
+        git clone https://github.com/mendersoftware/mender-artifact.git "${mendersoftware_dir}"/mender-artifact
+        git -C "${mendersoftware_dir}"/mender-artifact checkout "${MENDER_ARTIFACT_VER}"
+    popd
+
+    pushd "${TOOLSET_DIR}/mender/uboot-mender"
+        info "building Das U-Boot"
+
+        ARCH=arm CROSS_COMPILE="${cross_compiler}" make --quiet distclean
+        ARCH=arm CROSS_COMPILE="${cross_compiler}" make rpi_3_32b_defconfig
```
Comment (denismosolov) [inline toolset.sh]: Device specific

File: pieman.sh [hunk 30]
```diff
@@ -183,48 +228,114 @@ run_scripts "bootstrap"
...
+    cp "${TOOLSET_DIR}"/mender/u-boot.bin "${MOUNT_POINT}"/kernel7.img
+    sed -i -e "s/\b[ ]root=[^ ]*/ root=\${mender_kernel_root}/" "${MOUNT_POINT}"/cmdline.txt
+
+    umount "${MOUNT_POINT}"
+
+    mount "${LOOP_DEV}p2" "${MOUNT_POINT}"
+
+    rsync -apS "${R}"/ "${MOUNT_POINT}"
+
+    install_mender
```
Comment (denismosolov) [inline pieman.sh]: Let's install mender before rsync and before calculating image size

File: files/mender/boot.cmd [hunk 31]
```diff
@@ -0,0 +1,6 @@
+fdt addr ${fdt_addr} && fdt get value bootargs /chosen bootargs
+run mender_setup
+mmc dev ${mender_uboot_dev}
+load ${mender_uboot_root} ${kernel_addr_r} /boot/zImage
+bootz ${kernel_addr_r} - ${fdt_addr}
+run mender_try_to_recover
```
Comment (denismosolov) [inline files/mender/boot.cmd]: Device specific

**Pull request merged**